### DISCUSSION TITLE
License of exported FMU from OpenModelica

### DISCUSSION THREAD

**Discussion by Arinomo** [OpenModelica/OpenModelica#10929] [General]
Edit:
Say I made a model and export it as an FMU from OM. The FMU is then integrated in my application, that simulate said fmu. The application will than be used internally in my organization. Does it allowed by OpenModelica License? when yes, since it only be used internally, could the FMU contain source file or it should only be used as blackbox ([reference](https://openmodelica.org/forum/default-topic/2666-binary-executable-or-fmu-to-commercial-use.html))?

**Comment by casella** [comment]
Stuff compiled with OMC falls under the [OpenModelica Run-time System Public License](https://openmodelica.org/osmc-pl/osmc-pl-1.8-runtime.txt). BSD is very permissive, so I don't think you should have problems.

**Reply by Arinomo** [reply]
@casella thanks for the reply. Exported FMU includes header files in source that are not explicitly state use of BSD (such as `dassl.h`  for example), hence the question, could I include FMU with source file in my application.

If I understand correctly, CS-FMU use `Euler` and so the  `dassl.h` is not necessary for FMU simulation right?

**Reply by casella** [reply]
Maybe we are not using the right header files, would you mind opening a ticket about that?

I guess dassl.h is not needed for CS-FMU, though I'm not 100% sure. Maybe @mahge can comment on that.

**Reply by AnHeuermann** [reply]
I'm not a lawyer, so no legal advice:
Source-Code FMUs (a default option) export parts of the OpenModelica source code (duh!), so that OSMC-PL (AGPL) or GPL or BSD should apply.
In addition we include source files of some third-party tools. To name a few: CVODE from SUNDIALS, LAPACK and BLAS
So those licenses could apply as well.

FMUs that are linked statically have object code from OpenModelica and I think in that case the OSMC-PL (AGPL) should apply as well.
For the third party tools licenses might apply as well.

From the cvode_solver.h (shipped in CS FMUs when CVODE solver is chosen):
````
/*
 * This file is part of OpenModelica.
 *
 * Copyright (c) 1998-CurrentYear, Open Source Modelica Consortium (OSMC),
 * c/o Linköpings universitet, Department of Computer and Information Science,
 * SE-58183 Linköping, Sweden.
 *
 * All rights reserved.
 *
 * THIS PROGRAM IS PROVIDED UNDER THE TERMS OF THE BSD NEW LICENSE OR THE
 * GPL VERSION 3 LICENSE OR THE OSMC PUBLIC LICENSE (OSMC-PL) VERSION 1.2.
 * ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS PROGRAM CONSTITUTES
 * RECIPIENT'S ACCEPTANCE OF THE OSMC PUBLIC LICENSE OR THE GPL VERSION 3,
 * ACCORDING TO RECIPIENTS CHOICE.
 *
 * The OpenModelica software and the OSMC (Open Source Modelica Consortium)
 * Public License (OSMC-PL) are obtained from OSMC, either from the above
 * address, from the URLs: http://www.openmodelica.org or
 * http://www.ida.liu.se/projects/OpenModelica, and in the OpenModelica
 * distribution. GNU version 3 is obtained from:
 * http://www.gnu.org/copyleft/gpl.html. The New BSD License is obtained from:
 * http://www.opensource.org/licenses/BSD-3-Clause.
 *
 * This program is distributed WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, EXCEPT AS
 * EXPRESSLY SET FORTH IN THE BY RECIPIENT SELECTED SUBSIDIARY LICENSE
 * CONDITIONS OF OSMC-PL.
 *
 */
```

**Reply by Arinomo** [reply]
> object code from OpenModelica

@AnHeuermann can you elaborate, what is an object code in this contex? is it `fmu.dll` in the binary files?


> Source-Code FMUs (a default option) export parts of the OpenModelica source code (duh!), so that OSMC-PL (AGPL) or GPL or BSD should apply.

i've opened #10935 since there are some header files that are not explicitly state as BSD

**Reply by AnHeuermann** [reply]
I'm not sure about this, maybe https://www.gnu.org/licenses/gpl-faq.html#v3Under4and5 helps.

What object code is in and FMU?
Depends on your target system (Windows, Linux, ...) and the way you compiled the FMU.
For a random FMU I found
```
$ ls FMU/binaries/win64/Modelica_Blocks_Examples_PID_Controller*
FMU/binaries/win64/Modelica_Blocks_Examples_PID_Controller.dll  FMU/binaries/win64/Modelica_Blocks_Examples_PID_Controller.lib  FMU/binaries/win64/Modelica_Blocks_Examples_PID_Controller_FMU.libs
```
Here libs should be a static library.

And you can explicitly link statically when building a FMU:
```
buildMOdelFMU(Modelica.Blocks.Examples.PID_Controller, platforms={"static"});
```
(Actually the CMake FMUs ignore this and will prefer static libs as default).

**Reply by Arinomo** [reply]
> FMUs that are linked statically have object code from OpenModelica and I think in that case the OSMC-PL (AGPL) should apply as well.

> Here libs should be a static library.

> (Actually the CMake FMUs ignore this and will prefer static libs as default).

So if I export following model as blackbox 

```.mo
model foo
  parameter Real userGain=10;
  Modelica.Blocks.Interfaces.RealInput u annotation(
    Placement(visible = true, transformation(origin = {-82, 0}, extent = {{-20, -20}, {20, 20}}, rotation = 0), iconTransformation(origin = {-82, 0}, extent = {{-20, -20}, {20, 20}}, rotation = 0)));
  Modelica.Blocks.Interfaces.RealOutput y annotation(
    Placement(visible = true, transformation(origin = {78, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0), iconTransformation(origin = {78, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Blocks.Math.Gain gain(k = userGain)  annotation(
    Placement(visible = true, transformation(origin = {-4, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
equation
  connect(u, gain.u) annotation(
    Line(points = {{-82, 0}, {-16, 0}}, color = {0, 0, 127}));
  connect(gain.y, y) annotation(
    Line(points = {{8, 0}, {78, 0}}, color = {0, 0, 127}));

annotation(
    uses(Modelica(version = "4.0.0")));
end foo;
```
The FMU would be licensed under OSMC-PL(AGPL)? and subsequently, my application will also licensed as AGPL, i.e. i have to make the source code public?

**Reply by AnHeuermann** [reply]
Sorry, I'm not a lawyer. When in doubt contact a lawyer.
This stuff can be super complicated and you should never trust random people from the internet (even though they might be associated with the software).

  - https://www.gnu.org/licenses/gpl-faq.html#GPLRequireSourcePostedPublic
  - https://www.gnu.org/licenses/gpl-faq.html#GPLCommercially

**Reply by Arinomo** [reply]
@AnHeuermann will do. Thanks for the input.
